errors.go

```diff
@@ -25,387 +25,99 @@ package errdefs
-type dataLoss interface {
-	DataLoss()
-}
-
-// IsDataLoss returns true if data during an operation was lost or corrupted
-func IsDataLoss(err error) bool {
-	return errors.Is(err, errDataLoss{}) || isInterface[dataLoss](err)
-}
-
-type errUnauthorized struct{}
-
-func (errUnauthorized) Error() string { return "unauthorized" }
-
-func (errUnauthorized) Unauthorized() {}
-
-func (e errUnauthorized) WithMessage(msg string) error {
-	return customMessage{e, msg}
-}
-
-// unauthorized maps to Moby's "ErrUnauthorized"
-type unauthorized interface {
-	Unauthorized()
+var ErrCanceled = context.Canceled
+
+func (e Error) Error() string {
+	switch e {
+	case ErrInvalidArgument:
+		return "invalid argument"
+	case ErrNotFound:
+		return "not found"
+	case ErrAlreadyExists:
+		return "already exists"
+	case ErrPermissionDenied:
+		return "permission denied"
+	case ErrResourceExhausted:
+		return "resource exhausted"
+	case ErrFailedPrecondition:
+		return "failed precondition"
+	case ErrConflict:
+		return "conflict"
+	case ErrNotModified:
+		return "not modified"
+	case ErrAborted:
+		return "aborted"
+	case ErrOutOfRange:
+		return "out of range"
+	case ErrNotImplemented:
+		return "not implemented"
+	case ErrInternal:
+		return "internal"
+	case ErrUnavailable:
+		return "unavailable"
+	case ErrDataLoss:
+		return "unauthenticated"
+	case ErrUnauthenticated:
+		return "unauthenticated"
+	default:
+		return "unknown"
+	}
 }
 
-// IsUnauthorized returns true if the error indicates that the user was
-// unauthenticated or unauthorized.
-func IsUnauthorized(err error) bool {
-	return errors.Is(err, errUnauthorized{}) || isInterface[unauthorized](err)
+func (e Error) WithMessage(msg string) error {
+	return customMessage{
+		msg: msg,
+		err: e,
+	}
 }
 
-func isInterface[T any](err error) bool {
-	for {
-		switch x := err.(type) {
-		case T:
-			return true
-		case customMessage:
-			err = x.err
-		case interface{ Unwrap() error }:
-			err = x.Unwrap()
-			if err == nil {
-				return false
-			}
-		case interface{ Unwrap() []error }:
-			for _, err := range x.Unwrap() {
-				if isInterface[T](err) {
-					return true
-				}
-			}
-			return false
-		default:
-			return false
-		}
+func (e Error) WithMessagef(format string, args ...any) error {
+	return customMessage{
+		msg: fmt.Sprintf(format, args...),
+		err: e,
 	}
 }
 
-// customMessage is used to provide a defined error with a custom message.
-// The message is not wrapped but can be compared by the `Is(error) bool` interface.
+// customMessage wraps an underlying code with a custom message.
 type customMessage struct {
-	err error
 	msg string
+	err Error
 }
 
-func (c customMessage) Is(err error) bool {
-	return c.err == err
-}
-
-func (c customMessage) As(target any) bool {
-	return errors.As(c.err, target)
+func (m customMessage) Error() string {
+	return m.msg
 }
 
-func (c customMessage) Error() string {
-	return c.msg
+func (m customMessage) Unwrap() error {
```

Comment:
Why `Unwrap` vs using the `Is` and `As` or even keeping both?

Reply:
I honestly thought I was reverting this and thought this was the original code.

I do think `Unwrap()` is probably better since it causes the standard library to do things properly and makes it so there's fewer things to test. While it isn't in this PR, I think `customMessage` is likely not needed as its own type and could be replaced by `Attach` or `Mask` from https://github.com/containerd/errdefs/pull/16, but `Attach` and `Mask` provide a pattern that's more composable in other situations.

For example, say I want to annotate an existing error with an error from errdefs. I would either have to make another error type that wraps errdefs or I could utilize one of the functions in that PR and get the same behavior.